File: arrays_and_hashing/group_anagrams.py
# Q:49 - https://leetcode.com/problems/group-anagrams/

from typing import List
from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        map = defaultdict(list)
        for str in strs:
            charList = [0] * 26
            for ch in str:
                index = ord(ch) - ord("a")
                charList[index] += 1
            

            key = tuple(charList)

            map[key].append(str)

        return list(map.values())

c = Solution()

strs1 = ["eat","tea","tan","ate","nat","bat"]
strs2 = ["a"]

print(c.groupAnagrams(strs1))
print(c.groupAnagrams(strs2))

File: arrays_and_hashing/group_anagrams.js
// Q:49 - https://leetcode.com/problems/group-anagrams/

/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
  const map = {};

  for (let i = 0; i < strs.length; i++) {
    const currentStr = strs[i];

    const arr = Array(26).fill(0);

    for (let j = 0; j < currentStr.length; j++) {
      const currentChar = currentStr[j];
      const code = currentChar.charCodeAt(0) - "a".charCodeAt(0);
      arr[code] += 1;
    }

    arrKey = arr.join(",");

    if (!map.hasOwnProperty(arrKey)) {
      map[arrKey] = [];
    }

    map[arrKey].push(currentStr);
  }

  return Object.values(map);
};
